src/main/scala/calc/Compiler.scala

```diff
@@ -61,12 +75,49 @@ object Compiler {
     ir.Hashers.hashClassDef(classDef)
   }
 
+  def findFreeVar(tree: Tree, listId: Set[String]): Set[String]= {
+    implicit val pos = tree.pos
+
+    tree match {
+      case Literal(value) => Set()
+
+      case BinaryOp(op, rhs, lhs) =>
+        findFreeVar(rhs, listId) ++ findFreeVar(lhs, listId)
+
+      case Let(name, value, body) =>
+        findFreeVar(value, listId) ++ findFreeVar(body, listId + name.name)
+
+      case Ident(name) =>
+        if(listId.contains(name)) Set()
+        else Set(name)
+
+      case If(cond, thenp, elsep) =>
+        findFreeVar(cond, listId)  ++
+        findFreeVar(thenp, listId) ++
+        findFreeVar(elsep, listId)
+
+      case Closure(params, body) =>
+        val typePar = params map ( x => x.name )
+        findFreeVar(body, listId ++ typePar)
+
+      case Call(Ident(name) , args)
+        if (!(listId.contains(name)) && nativeFun.contains(name)) =>
+         args.map( x => findFreeVar(x, listId) ).reduce( (x,y) => x++y ).toSet
+
+      case Call(fun , args) =>
+        val argSet = args.map( x => findFreeVar(x, listId) )
+          .reduce( (x, y) => x ++ y )
+          .toSet
```

Comment:
There's `flatMap` for that: `args.flatMap(x => findFreeVar(x, listId))`. (same for the other case)